common/txmgr/address_state.go

```diff
@@ -0,0 +1,697 @@
+		TxFatalError:              0,
+	}
+	for _, tx := range txs {
+		counts[tx.State]++
+	}
+
+	as := AddressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]{
+		lggr:        lggr,
+		chainID:     chainID,
+		fromAddress: fromAddress,
+
+		idempotencyKeyToTx:      make(map[string]*txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE], len(txs)),
+		unstarted:               NewTxPriorityQueue[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE](maxUnstarted),
+		inprogress:              nil,
+		unconfirmed:             make(map[int64]*txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE], counts[TxUnconfirmed]),
+		confirmedMissingReceipt: make(map[int64]*txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE], counts[TxConfirmedMissingReceipt]),
+		confirmed:               make(map[int64]*txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE], counts[TxConfirmed]),
+		allTransactions:         make(map[int64]*txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE], len(txs)),
+		fatalErrored:            make(map[int64]*txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE], counts[TxFatalError]),
+	}
+
+	// Load all transactions supplied
+	for i := 0; i < len(txs); i++ {
+		tx := txs[i]
+		switch tx.State {
+		case TxUnstarted:
+			as.unstarted.AddTx(&tx)
+		case TxInProgress:
+			as.inprogress = &tx
+		case TxUnconfirmed:
+			as.unconfirmed[tx.ID] = &tx
+		case TxConfirmedMissingReceipt:
+			as.confirmedMissingReceipt[tx.ID] = &tx
+		case TxConfirmed:
+			as.confirmed[tx.ID] = &tx
+		case TxFatalError:
+			as.fatalErrored[tx.ID] = &tx
+		}
+		as.allTransactions[tx.ID] = &tx
+		if tx.IdempotencyKey != nil {
+			as.idempotencyKeyToTx[*tx.IdempotencyKey] = &tx
+		}
+	}
+
+	return &as, nil
+}
+
+func (as *AddressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) close() {
+	clear(as.idempotencyKeyToTx)
+
+	as.unstarted.Close()
+	as.unstarted = nil
+	as.inprogress = nil
+
+	clear(as.unconfirmed)
+	clear(as.confirmedMissingReceipt)
+	clear(as.confirmed)
+	clear(as.allTransactions)
+	clear(as.fatalErrored)
+
+	as.idempotencyKeyToTx = nil
+	as.unconfirmed = nil
+	as.confirmedMissingReceipt = nil
+	as.confirmed = nil
+	as.allTransactions = nil
+	as.fatalErrored = nil
+}
+
+func (as *AddressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) CountTransactionsByState(txState txmgrtypes.TxState) int {
+	as.RLock()
+	defer as.RUnlock()
+
+	switch txState {
+	case TxUnstarted:
+		return as.unstarted.Len()
+	case TxInProgress:
+		if as.inprogress != nil {
+			return 1
+		}
+	case TxUnconfirmed:
+		return len(as.unconfirmed)
+	case TxConfirmedMissingReceipt:
+		return len(as.confirmedMissingReceipt)
+	case TxConfirmed:
+		return len(as.confirmed)
+	case TxFatalError:
+		return len(as.fatalErrored)
+	}
+
+	return 0
+}
+
+func (as *AddressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) FindTxWithIdempotencyKey(key string) *txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE] {
```

Comment:
I think I remember Dmytro bringing this up but shouldn't we avoid returning a pointer to the tx here?

Reply:
Disregard this. I hadn't made it down to the inmemory store code where you deep copy